doc/dev/schema.md

```diff
@@ -0,0 +1,193 @@
+is obtained, used and reloaded.
+
+## Table of Contents
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Space schema](#space-schema)
+  - [How schema is stored](#how-schema-is-stored)
+  - [When schema is used](#when-schema-is-used)
+  - [How schema is reloaded](#how-schema-is-reloaded)
+  - [When schema is reloaded and operation is retried](#when-schema-is-reloaded-and-operation-is-retried)
+  - [Alternative approaches](#alternative-approaches)
+- [Sharding schema](#sharding-schema)
+  - [How schema is stored](#how-schema-is-stored-1)
+  - [When schema is used](#when-schema-is-used-1)
+  - [How schema is reloaded](#how-schema-is-reloaded-1)
+  - [When schema is reloaded and operation is retried](#when-schema-is-reloaded-and-operation-is-retried-1)
+  - [Alternative approaches](#alternative-approaches-1)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Space schema
+
+Related links: [#98](https://github.com/tarantool/crud/issues/98),
+[PR#111](https://github.com/tarantool/crud/pull/111).
+
+### How schema is stored
+
+Every ``crud`` router is a [``vshard``](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/)
+router, the same applies to storages. In ``vshard`` clusters, spaces are
+created on storages. Thus, each storage has a schema (space list,
+space formats and indexes) on it.
+
+Every router has a [``net.box``](https://www.tarantool.io/en/doc/latest/reference/reference_lua/net_box/)
+connection to each storage it could interact with. (They can be
+retrieved with [``vshard.router.routeall``](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_router/#lua-function.vshard.router.routeall)
+call.) Each ``net.box`` connection has space schema for an instance it
+is connected to. Router can access space schema by using ``net.box``
+connection object contents.
+
+### When schema is used
+
+Space schema is used to flatten maps on `crud.*_object*` requests.
+Space schema is used to resolve update operation fields if updating
+by field is not supported natively. Together with sharding info,
+the space schema is used to calculate ``bucket_id`` to choose
+replicaset for a request. Finally, schema is returned to a user
+in `metadata` field of request response, so it could be used later
+for `crud.unflatten_rows`.
+
+### How schema is reloaded
+
+``net.box`` schema reload works as follows. For each request (we use ``call``s
+to execute our procedures) from client (router) to server (storage) we
+receive a schema version together with response data. If schema versions mismatch,
+client reloads schema. The request is not retried. ``net.box`` reloads
+schema before returning ``call`` result to a user, so the next request will use
+updated schema. (See [tarantool/tarantool/6169](https://github.com/tarantool/tarantool/issues/6169)).
```

Comment:
1. Next requests *in this fiber*.
2. If the request is synchronous.

Both points are my guesses about the implementation (it is likely so, but I don't check it specifically).

Reply:
But next request in another fiber will also use updated schema since there is the only one net.box schema object.

Reply:
Here you describe `net.box` behaviour, so it would be good it make this description correct and complete.

Reply:
So if we use asynchronous request, it doesn't reload the schema if user fetches the result?

Reply:
I'll cite current text:

> ``net.box`` reloads schema before returning ``call`` result to a user, so the next request will use updated schema.

Let's consider an example:

```lua
conn:call('foo', {<..foo args..>}, {is_async = true})
local result = conn.space.s:select('bar', {<..select args..>})
```

According to your text if calling 'foo' reloads the schema, we'll see its effect in the select (because the `call` returns at this point). I doubt that it is guaranteed (but I didn't check the net.box's code).

Reply:
I guess that sync request works in the same way as async request + `:wait_result()`.

Reply:
> 

well

> I'll cite current text:
> 
> > `net.box` reloads schema before returning `call` result to a user, so the next request will use updated schema.
> 
> Let's consider an example:
> 
> ```lua
> conn:call('foo', {<..foo args..>}, {is_async = true})
> local result = conn.space.s:select('bar', {<..select args..>})
> ```
> 
> According to your text if calling 'foo' reloads the schema, we'll see its effect in the select (because the `call` returns at this point). I doubt that it is guaranteed (but I didn't check the net.box's code).

Well, it seems that I need to replace `result` with something like `synchronous result`

Reply:
I've tried to improve the text based on this discussion.

Reply:
Thank you!

> so the next request in a fiber will use updated schema

I think it should be:

> so **a** next request in **the** fiber will use **the** updated schema